tests/tests/dos.rs

```diff
@@ -378,3 +385,91 @@ async fn complex_queries__2_state_transition_bytecode__query_to_complex() {
 }
+
+#[tokio::test]
+async fn concurrency_limit_0_prevents_any_queries() {
+    // Given
+    let mut config = Config::local_node();
+    config.graphql_config.max_concurrent_queries = 0;
+    config.graphql_config.api_request_timeout = std::time::Duration::from_millis(100);
+
+    let query = FULL_BLOCK_QUERY.to_string();
+    let query = query.replace("$NUMBER_OF_BLOCKS", "40");
+
+    let node = FuelService::new_node(config).await.unwrap();
+    let url = format!("http://{}/v1/graphql", node.bound_address);
+    let client = reqwest::Client::new();
+
+    // When
+    let response = client
+        .post(url)
+        .header("Content-Type", "application/json")
+        .header("Content-Length", "10_000")
+        .body(vec![42; 10_000])
+        .send()
+        .await;
+
+    // Then
+    let result = response.unwrap();
+    assert_eq!(result.status(), 408);
+}
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 8)]
+async fn concurrency_limit_1_prevents_concurrent_queries() {
+    // Given
+    let num_samples = 100;
+
+    let mut config = Config::local_node();
+    config.graphql_config.max_concurrent_queries = 1;
+
+    let query = FULL_BLOCK_QUERY.to_string();
+    let query = query.replace("$NUMBER_OF_BLOCKS", "40");
+
+    let node = FuelService::new_node(config).await.unwrap();
+    let url = format!("http://{}/v1/graphql", node.bound_address);
+    let client = reqwest::Client::new();
+
+    let (mut tx, mut rx) = tokio::sync::mpsc::channel(100);
+
+    // When
+
+    // Measure the average request time for sequential queries
+    let mut avg_request_time = 0;
+    for _ in 0..num_samples {
+        let now = Instant::now();
+        send_graph_ql_query(&url, &query).await;
+        avg_request_time += now.elapsed().as_nanos() / num_samples;
+    }
+
+    // Measure the average request time for concurrent queries
+    for idx in 0..num_samples {
+        let tx = tx.clone();
+        let url = url.clone();
+        let query = query.clone();
+
+        tokio::spawn(async move {
+            let now = Instant::now();
+            send_graph_ql_query(&url, &query).await;
+            let _ = tx.send(now.elapsed().as_nanos()).await;
+        });
+    }
+
+    let mut avg_concurrent_request_time = 0;
+    for idx in 0..num_samples {
+        let request_time = rx.recv().await.unwrap();
+        avg_concurrent_request_time += request_time / num_samples;
+    }
+
+    // Then
+
+    // In an idealized model we should see c = s * n / 2
+    // where
+    //   c = average concurrent request time
+    //   s = single request time
+    //   n = number of request
+    //   2 = the first even natural non-zero number ;)
+    //
+    // However, since this is inherently flaky we divide by 4 instead of 2 to have some margin,
+    // while still maintaining our ability to assert a large deviation between the two measurements.
+    assert!(avg_concurrent_request_time > avg_request_time * num_samples / 4);
```

Comment:
Maybe we want to increase number of samples to be `10_000` to be sure that values will be big and remove any divide

Reply:
10_000 takes a long time. The test already takes a couple of seconds with number of samples being `100`. Though I checked now for a couple of runs and the deviation between the average concurrent request time and the average serial request time multiplied by the number of samples divided by two (s * n / 2) never deviates more than one percent. 

I could instead make an assertion about the relative difference between these two.

Reply:
I also use nanoseconds here to reduce the impact of integer division errors.

Reply:
Actually using relative difference has more problems in case we get extra delays not accounted for in the model etc. I'd like to stick with the simple equality check for now as the current setup does these two things:

1. It proves that we get a significant slowdown with the concurrency limit in place. This assertion would fail if we don't see a huge slowdown in the same order of magnitude as the expected slowdown caused by the concurrency limit.
2. It has enough margin to hopefully not become one of those annoying flaky tests which occasionally fails in CI.